### PR TITLE
Add: Help and manuals window

### PR DESCRIPTION
**Pull request by nielsmh**
image

This adds another button to the main menu, letting you access both included documentation in the package, as well as links to various official online resources. The Markdown format files are parsed, offering a Table of Contents dropdown and working hyperlinks.
It also adds this choice to the in-game help/about menu.

Several issues:
- [x] The changelog file is too long for the text file viewer. Currently scrollbars are limited to 65535 items, so this should be expanded. Probably best handled by a separate change/fix.
- [x] Markdown does not look very nice in the viewer. Can be solved by making the viewer process markdown to make it look better, or by shipping processed files with the game.
- [x] The location of documentation files is not entirely reliable, especially on Unix and Linux systems, the packaging has several places it could put them (though typically under /usr/share somewhere.) The file searching needs to be improved.
- [x] For some reason the COPYING file (with no extension) does not get displayed, even though my debugger says it is being read into memory.
- [x] It seems to be an open question whether TT-Forums is sufficiently official to warrant being linked from inside the game.
- [x] The Markdown parsing should be moved out of the Help window so it can be used for user content as well.

### CODE REVIEW

**Comment by James103** [comment]
> Currently scrollbars are limited to 65535 items

This is probably why the game crashes when you generate too many news messages at once (#7613).

**Comment by nielsmh** [comment]
Some fancyness: Internal anchor links are green, and web links are blue. Both work, though at most one link per (source file) line.

image

Possibly also to do: Make links to other files work, and add a back button.

**Comment by nielsmh** [comment]
Does anyone have suggestions for how to best handle the "finding the right path for the docs files" problem? In particular on Unix-y systems where the docs may be in /usr/share/openttd while the binary may be in /usr/bin. Are any of the existing search paths useful for this, or is a new one necessary?

**Comment by James103** [comment]
```
Commit checker: Expected — Waiting for status to be reported (Required)
```
a) Can you please re-run the checks, preferably by rebasing this PR?

**Comment by nielsmh** [comment]
Can maybe make more of the changelog viewable with the scrollbar fix in #8006 too. Although the text file viewer is already slow with many lines.

**Comment by nielsmh** [comment]
@TrueBrain mentioned that he'd like if the Markdown parsing was moved around, so it could also be used for user content and basesets.

There's also a question of where to put the feature on the main menu. Is it really the right thing to do, to remove the highscore table button?

**Comment by LordAro** [comment]
I know I've never (intentionally) clicked it :)
TBH, I'm not sure anyone would much mind if the highscore table was removed entirely...

**Comment by sheepo99** [comment]
Linking to the manual is nice, but I do not see the purpose for all the other buttons. Why providing players with links to resources mostly aimed at developers who would end up consulting the documentation with the same info either way? This makes no sense from a user experience POV, much less as a replacement for the highscore board, which is an actual user experience feature.

**Comment by nielsmh** [comment]
Which of my proposed buttons are mostly aimed at developers?
- The readme file is intended for end users, it contains various installation information.
- The changelog is intended for end users, to know what changed between versions.
- The known bugs file is intended for end users, to learn about standing (and possibly unfixable) issues.
- The license is actually required to be viewable from within the program according to the terms of the license. And it is intended for end users to know their rights regarding copying the program and obtaining the source code.
- Official website, official wiki, community forum, are intended for end users, to know where the actual developers reside and what the official source for the game is. (If gaming magazine cover CDs were still a thing, they would have even more raison d´etrê.)
- Report a bug is there so end users can know where the actual place to report bugs is, and protect themselves and us from discussing issues in places none of us would learn about it.
- Community forums is there to point to the traditional home.

As for replacing the high score screen, nobody ever said that was a final decision. However the high score game mechanic makes little sense today, since the scoring is rather arbitrary, and does not take difficulty, game settings, or mods into account. The entire main menu experience is kind of weird considering all the extensions and changes to gameplay, e.g. the four big landscape selection buttons pertain to new games only but are front and center. A completely redesigned main menu could offer a much better location for a help and manuals button.

**Comment by sheepo99** [comment]
I completely agree that the main menu requires a redesign, and if modding does indeed make the scoreboard irrelevant, then it should go. My proposition for the time being, however, would be to add all the relevant info to the About window (which already contains the link to the website), and add a new button for it either from the Options window, or the main menu. This is for example what OpenRCT2 does:

image

**Comment by nielsmh** [comment]
Well here we go, I did manage to make multiple hyperlinks in one line work. Sort of.
There is something weird with how I count characters versus how the text layouter counts characters (for the purpose of getting "character at X pixel position") that makes it miscount in some places.

**Comment by nielsmh** [comment]
Crash and not-working-at-all (with sprite font) issues with links fixed, but still have an outstanding issue with how the click position on the line is measured, causing links to extend the click position for all following text. Right now it's "kind of acceptable", but it might become a problem in the future and probably better to fix now.

Still need to make a plan for squashing.

**Comment by nielsmh** [comment]
There we go, squashed and neat to look at now.

**Comment by nielsmh** [comment]
The finding files and browsing around in them works on Windows and Emscripten, but no idea about Mac or Unix builds. Someone needs to test those.

**Comment by nielsmh** [comment]
In today's rebase I removed a commit that was supposed to make text directionality carry between files and possibly be deduced by the language of the file, but it seemed that commit was incomplete and didn't really do anything useful at all. The text file reader probably won't handle text with default directionality opposite of the UI language correctly.

File: src/help_gui.cpp [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+
+
+static const char README_FILENAME[] = "README.md";
+static const char CHANGELOG_FILENAME[] = "changelog.txt";
+static const char KNOWN_BUGS_FILENAME[] = "known-bugs.txt";
+static const char LICENSE_FILENAME[] = "COPYING.md";
+
+static std::string FindGameManualFilePath(const char *filename)
+{
+	/* FIXME: this is very arbitrary, should perhaps add a dedicated documentation searchpath? */
```
[the remark this answers is not in the document]
Comment (nielsmh) [reply src/help_gui.cpp]: This is the only open issue with this now, as I see it. I don't know if what I'm doing is correct, or at least "good enough". Can anyone help shed light on this?

**Comment by nielsmh** [comment]
I just realized that the `docs/` folder does not get installed currently. That means that even if README.md does get installed and found, it will have a number of dead links in it.
So that leaves three options:
1. Leave the `docs/´ out but keep README.md in, with broken links.
2. Drop offering the Readme for viewing from in-game.
3. Add `docs/` to the installation, or at least the text files parts.

**Comment by PeterN** [comment]
Are you able to split off the layouter fixes into (a) separate PR(s) as appropriate? On the principle of separation-of-concerns.

**Comment by 2TallTyler** [comment]
`Change: Add more hyperlinks within the included documentation` can definitely be a separate PR too, since those files can also be browsed through GitHub.

I'd very much like to see this PR go in, but lack the expertise to review further.

File: src/help_gui.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (TrueBrain) [reply src/help_gui.cpp]: This is a tricky question. On Linux, any decent distro should be setting SP_SHARED_DIR to point to `/usr/share/openttd`, or whatever they are using. So I think we can leave it up to them to do this properly, honestly.
For our distribution of Linux, which is a tarball extracted, SP_BINARY_DIR will do fine.

On Windows, most people install it via the .zip (so via SP_BINARY_DIR) and via Steam it is the same.

On MacOS, it is in the app-bundle, so there it should be valid too.

So I think you covered them all; will do some testing to actually confirm over all the different OSes and our distribution platforms, but yeah, we should be good to go here.

File: src/widgets/help_widget.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
+/* $Id$ */
```
Comment (PeterN) [inline src/widgets/help_widget.h]: Hello svn

**Comment by TrueBrain** [comment]
Right, after asking @nielsmh decided to pick this PR up again. It really was nearly done. I moved the requested commit to #11288 , and did a bunch of cleanup / update to our latest state of code (read: removing C-strings and using the C++ variants we have now). I kept that in a separate commit, so it is more visible what I changed. We will squash this PR when it is done I assumed anyway.

The quality of this Pull Request is REALLY high. There is so much stuff in here .. working links, history, table-of-content .. you did an amazing job @nielsmh , it is really lovely to work with a Pull Request of this quality. So my job was also really easy :)

Next up, is some testing. I will build some release binaries based on this work, and test in various of OSes if it is actually working. To keep a list for myself:

- [ ] Test Linux Generic
- [ ] Test Windows
- [ ] Test MacOS app-bundle

**Comment by TrueBrain** [comment]
All files can be found on all OSes we support via our distribution mechanisms. So we should be good there. Found a few bugs that I am addressing (this to keep track of them):

- [ ] Clicking on a directory crashes. `fopen` succeeds on a directory, with a very very large filesize.
- [ ] Table of Content isn't really visible.
- [ ] We should ask a user before we open a browser, as user-generated content can also contain links.

**Review by TrueBrain — APPROVED** [review]

**Pull request merged**